File: hosts/rust/src/transport/websocket_server.rs
// Quackoscope host (Rust) -- transport layer.
//
// An RFC 6455 WebSocket server over std::net, one OS thread per connection.
// It is written out here rather than pulled from a crate so that the whole
// host builds from two dependencies -- the openDAQ bindings and serde_json --
// and so that the framing the frontend sees is visible in this repository.
//
// Nothing in this file mentions the openDAQ crate or the M1 service DTOs: it
// speaks text frames, binary frames and the transport::wire envelopes only.

use std::collections::HashMap;
use std::io::{BufRead, BufReader, Read, Write};
use std::net::{Shutdown, TcpListener, TcpStream};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};

use serde_json::Value as Json;

use super::wire::{self, Outcome};

/// The GUID RFC 6455 section 1.3 appends to Sec-WebSocket-Key before hashing.
const WEBSOCKET_ACCEPT_GUID: &str = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

/// Largest inbound message this host will assemble, in bytes. A control-plane
/// request is a few hundred bytes; anything past this is a client defect or an
/// attack, and the socket is closed rather than grown.
const MAX_INBOUND_MESSAGE_BYTES: usize = 1 << 20;

/// What the service layer implements so the transport can drive it.
pub trait ConnectionHandler: Send + Sync {
    fn on_open(&self, connection: &Arc<Connection>);
    fn on_close(&self, connection: &Arc<Connection>);
    fn on_request(&self, connection: &Arc<Connection>, method: &str, params: &Json) -> Outcome;
}

/// One live WebSocket. Cloneable across threads: the subscription pump threads
/// hold an `Arc<Connection>` and push binary frames through the same writer the
/// request thread uses, serialised by `writer`.
pub struct Connection {
    id: u64,
    peer: String,
    writer: Mutex<TcpStream>,
    open: AtomicBool,
}

impl Connection {
    pub fn id(&self) -> u64 {
        self.id
    }

    pub fn peer(&self) -> &str {
        &self.peer
    }

    pub fn is_open(&self) -> bool {
        self.open.load(Ordering::SeqCst)
    }

    /// Send one text frame (opcode 0x1).
    pub fn send_text(&self, text: &str) {
        self.send_frame(0x1, text.as_bytes());
    }

    /// Send one binary frame (opcode 0x2).
    pub fn send_binary(&self, bytes: &[u8]) {
        self.send_frame(0x2, bytes);
    }

    fn send_close(&self) {
        // 1000 = normal closure, RFC 6455 section 7.4.1.
        self.send_frame(0x8, &1000u16.to_be_bytes());
    }

    fn send_frame(&self, opcode: u8, payload: &[u8]) {
        if !self.open.load(Ordering::SeqCst) {
            return;
        }

        let mut frame = Vec::with_capacity(payload.len() + 10);
        frame.push(0x80 | opcode); // FIN set, no extensions, no fragmentation
        // Server-to-client frames are never masked (RFC 6455 section 5.1).
        if payload.len() < 126 {
            frame.push(payload.len() as u8);
        } else if payload.len() <= u16::MAX as usize {
            frame.push(126);
            frame.extend_from_slice(&(payload.len() as u16).to_be_bytes());
        } else {
            frame.push(127);
            frame.extend_from_slice(&(payload.len() as u64).to_be_bytes());
        }
        frame.extend_from_slice(payload);

        let mut writer = match self.writer.lock() {
            Ok(writer) => writer,
            Err(poisoned) => poisoned.into_inner(),
        };
        if let Err(e) = writer.write_all(&frame).and_then(|()| writer.flush()) {
            self.open.store(false, Ordering::SeqCst);
            eprintln!(
                "[transport] connection {} ({}): writing a {}-byte opcode-0x{opcode:X} frame failed: {e}",
                self.id,
                self.peer,
                payload.len()
            );
        }
    }
}

/// A parsed inbound frame header plus its unmasked payload.
struct InboundFrame {
    fin: bool,
    opcode: u8,
    payload: Vec<u8>,
}

pub struct WebSocketServer {
    listener: TcpListener,
    websocket_path: String,
    handler: Arc<dyn ConnectionHandler>,
    next_connection_id: AtomicU64,
}

impl WebSocketServer {
    /// Bind the listening socket now, so a port clash is reported before the
    /// host announces that it is listening.
    pub fn bind(
        address: &str,
        port: u16,
        websocket_path: &str,
        handler: Arc<dyn ConnectionHandler>,
    ) -> std::io::Result<WebSocketServer> {
        let listener = TcpListener::bind((address, port))?;
        Ok(WebSocketServer {
            listener,
            websocket_path: websocket_path.to_string(),
            handler,
            next_connection_id: AtomicU64::new(1),
        })
    }

    pub fn local_address(&self) -> std::io::Result<std::net::SocketAddr> {
        self.listener.local_addr()
    }

    /// Accept forever, one thread per connection.
    pub fn serve_until_process_exit(self) -> ! {
        let server = Arc::new(self);
        loop {
            match server.listener.accept() {
                Ok((stream, peer)) => {
                    let server = Arc::clone(&server);
                    let id = server.next_connection_id.fetch_add(1, Ordering::SeqCst);
                    std::thread::Builder::new()
                        .name(format!("quackoscope-websocket-connection-{id}"))
                        .spawn(move || server.run_connection(id, stream, peer.to_string()))
                        .expect("spawning a WebSocket connection thread");
                }
                Err(e) => eprintln!("[transport] accept() on the listening socket failed: {e}"),
            }
        }
    }

    fn run_connection(&self, id: u64, stream: TcpStream, peer: String) {
        let _ = stream.set_nodelay(true);

        let write_half = match stream.try_clone() {
            Ok(clone) => clone,
            Err(e) => {
                eprintln!("[transport] connection {id} ({peer}): could not clone the socket for writing: {e}");
                return;
            }
        };
        let mut reader = BufReader::new(stream);

        let request_line_and_headers = match read_http_request_head(&mut reader) {
            Ok(head) => head,
            Err(e) => {
                eprintln!("[transport] connection {id} ({peer}): reading the HTTP request head failed: {e}");
                return;
            }
        };

        let mut write_half = write_half;
        let key = match self.check_upgrade(&request_line_and_headers) {
            Ok(key) => key,
            Err(refusal) => {
                eprintln!("[transport] connection {id} ({peer}): {refusal}");
                let body = format!("{refusal}\n");
                let response = format!(
                    "HTTP/1.1 400 Bad Request\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{body}",
                    body.len()
                );
                let _ = write_half.write_all(response.as_bytes());
                let _ = write_half.flush();
                let _ = write_half.shutdown(Shutdown::Both);
                return;
            }
        };

        let accept = websocket_accept_value(&key);
        let response = format!(
            "HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: {accept}\r\n\r\n"
        );
        if let Err(e) = write_half
            .write_all(response.as_bytes())
            .and_then(|()| write_half.flush())
        {
            eprintln!("[transport] connection {id} ({peer}): writing the 101 response failed: {e}");
            return;
        }

        println!(
            "[transport] connection {id} ({peer}): upgraded to WebSocket on {}; Sec-WebSocket-Key {key} -> Sec-WebSocket-Accept {accept}",
            self.websocket_path
        );

        let connection = Arc::new(Connection {
            id,
            peer: peer.clone(),
            writer: Mutex::new(write_half),
            open: AtomicBool::new(true),
        });

        self.handler.on_open(&connection);
        self.pump_frames(&connection, &mut reader);
        connection.open.store(false, Ordering::SeqCst);
        self.handler.on_close(&connection);

        let closing = match connection.writer.lock() {
            Ok(writer) => writer.try_clone(),
            Err(poisoned) => poisoned.into_inner().try_clone(),
        };
        if let Ok(socket) = closing {
            let _ = socket.shutdown(Shutdown::Both);
        }
        println!("[transport] connection {id} ({peer}): closed");
    }

    /// The upgrade preconditions of RFC 6455 section 4.2.1, plus the path this
    /// host serves. On refusal the message names exactly what was wrong.
    fn check_upgrade(&self, head: &HttpRequestHead) -> Result<String, String> {
        if head.method != "GET" {
            return Err(format!(
                "the WebSocket upgrade needs a GET; the request line was \"{} {} {}\"",
                head.method, head.target, head.version
            ));
        }
        let path = head.target.split('?').next().unwrap_or(&head.target);
        if path != self.websocket_path {
            return Err(format!(
                "this host serves the WebSocket at {} only; the request asked for \"{}\"",
                self.websocket_path, head.target
            ));
        }
        match head.headers.get("upgrade") {
            Some(value) if value.eq_ignore_ascii_case("websocket") => {}
            Some(value) => return Err(format!("Upgrade header is \"{value}\", not \"websocket\"")),
            None => return Err("no Upgrade header in the request".to_string()),
        }
        match head.headers.get("sec-websocket-version") {
            Some(value) if value.trim() == "13" => {}
            Some(value) => {
                return Err(format!(
                    "Sec-WebSocket-Version is \"{value}\"; this host implements RFC 6455 version 13 only"
                ))
            }
            None => return Err("no Sec-WebSocket-Version header in the request".to_string()),
        }
        match head.headers.get("sec-websocket-key") {
            Some(key) if !key.trim().is_empty() => Ok(key.trim().to_string()),
            _ => Err("no Sec-WebSocket-Key header in the request".to_string()),
        }
    }

    /// Read frames, reassemble fragmented messages, answer control frames, and
    /// hand every complete text message to the handler as a wire request.
    fn pump_frames(&self, connection: &Arc<Connection>, reader: &mut BufReader<TcpStream>) {
        let mut message = Vec::new();
        let mut message_opcode: u8 = 0;

        loop {
            let frame = match read_frame(reader) {
                Ok(Some(frame)) => frame,
                Ok(None) => return, // peer closed the TCP connection
                Err(e) => {
                    eprintln!(
                        "[transport] connection {} ({}): reading a frame failed: {e}",
                        connection.id, connection.peer
                    );
                    return;
                }
            };

            match frame.opcode {
                0x8 => {
                    connection.send_close();
                    return;
                }
                0x9 => {
                    connection.send_frame(0xA, &frame.payload); // pong echoes the ping body
                    continue;
                }
                0xA => continue, // an unsolicited pong; nothing to do
                0x0 | 0x1 | 0x2 => {}
                other => {
                    eprintln!(
                        "[transport] connection {} ({}): opcode 0x{other:X} is not one this host handles; closing",
                        connection.id, connection.peer
                    );
                    connection.send_close();
                    return;
                }
            }

            if frame.opcode != 0x0 {
                message.clear();
                message_opcode = frame.opcode;
            }
            if message.len() + frame.payload.len() > MAX_INBOUND_MESSAGE_BYTES {
                eprintln!(
                    "[transport] connection {} ({}): an inbound message passed {MAX_INBOUND_MESSAGE_BYTES} bytes; closing",
                    connection.id, connection.peer
                );
                connection.send_close();
                return;
            }
            message.extend_from_slice(&frame.payload);

            if !frame.fin {
                continue;
            }

            if message_opcode == 0x2 {
                eprintln!(
                    "[transport] connection {} ({}): dropping a {}-byte client binary frame; the M1 contract's binary plane is server-to-client only",
                    connection.id,
                    connection.peer,
                    message.len()
                );
                message.clear();
                continue;
            }

            let text = match String::from_utf8(std::mem::take(&mut message)) {
                Ok(text) => text,
                Err(e) => {
                    eprintln!(
                        "[transport] connection {} ({}): a text frame was not valid UTF-8: {e}",
                        connection.id, connection.peer
                    );
                    connection.send_close();
                    return;
                }
            };

            self.answer_one_request(connection, &text);
        }
    }

    fn answer_one_request(&self, connection: &Arc<Connection>, text: &str) {
        let request = match wire::decode_request(text) {
            Ok(request) => request,
            Err(error_envelope) => {
                connection.send_text(&error_envelope.to_string());
                return;
            }
        };

        let outcome = self
            .handler
            .on_request(connection, &request.method, &request.params);

        let envelope = match outcome {
            Outcome::Result(result) => wire::encode_result(request.id, result),
            Outcome::Failure { code, detail } => wire::encode_error(request.id, &code, &detail),
        };
        connection.send_text(&envelope.to_string());
    }
}

struct HttpRequestHead {
    method: String,
    target: String,
    version: String,
    /// Header names lowercased; values trimmed.
    headers: HashMap<String, String>,
}

fn read_http_request_head(reader: &mut BufReader<TcpStream>) -> std::io::Result<HttpRequestHead> {
    let mut request_line = String::new();
    if reader.read_line(&mut request_line)? == 0 {
        return Err(std::io::Error::new(
            std::io::ErrorKind::UnexpectedEof,
            "the peer closed the socket before sending a request line",
        ));
    }
    let mut parts = request_line.trim_end().split_whitespace();
    let method = parts.next().unwrap_or_default().to_string();
    let target = parts.next().unwrap_or_default().to_string();
    let version = parts.next().unwrap_or_default().to_string();

    let mut headers = HashMap::new();
    loop {
        let mut line = String::new();
        if reader.read_line(&mut line)? == 0 {
            return Err(std::io::Error::new(
                std::io::ErrorKind::UnexpectedEof,
                "the peer closed the socket in the middle of the request headers",
            ));
        }
        let line = line.trim_end_matches(['\r', '\n']);
        if line.is_empty() {
            break;
        }
        if let Some((name, value)) = line.split_once(':') {
            headers.insert(name.trim().to_ascii_lowercase(), value.trim().to_string());
        }
    }

    Ok(HttpRequestHead {
        method,
        target,
        version,
        headers,
    })
}

/// One RFC 6455 frame. `Ok(None)` means the peer closed the TCP connection
/// cleanly between frames.
fn read_frame(reader: &mut BufReader<TcpStream>) -> std::io::Result<Option<InboundFrame>> {
    let mut first_two = [0u8; 2];
    match reader.read_exact(&mut first_two) {
        Ok(()) => {}
        Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(e) => return Err(e),
    }

    let fin = first_two[0] & 0x80 != 0;
    let opcode = first_two[0] & 0x0F;
    let masked = first_two[1] & 0x80 != 0;
    let short_length = (first_two[1] & 0x7F) as usize;

    let length = match short_length {
        126 => {
            let mut extended = [0u8; 2];
            reader.read_exact(&mut extended)?;
            u16::from_be_bytes(extended) as usize
        }
        127 => {
            let mut extended = [0u8; 8];
            reader.read_exact(&mut extended)?;
            let length = u64::from_be_bytes(extended);
            if length > MAX_INBOUND_MESSAGE_BYTES as u64 {
                return Err(std::io::Error::new(
                    std::io::ErrorKind::InvalidData,
                    format!(
                        "a frame announced {length} payload bytes, past this host's {MAX_INBOUND_MESSAGE_BYTES}-byte limit"
                    ),
                ));
            }
            length as usize
        }
        other => other,
    };

    if !masked {
        // RFC 6455 section 5.1: every client-to-server frame is masked.
        return Err(std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            format!("an inbound opcode-0x{opcode:X} frame of {length} bytes carried no mask"),
        ));
    }

    let mut mask = [0u8; 4];
    reader.read_exact(&mut mask)?;

    let mut payload = vec![0u8; length];
    reader.read_exact(&mut payload)?;
    for (index, byte) in payload.iter_mut().enumerate() {
        *byte ^= mask[index % 4];
    }

    Ok(Some(InboundFrame {
        fin,
        opcode,
        payload,
    }))
}

/// base64(SHA-1(key + GUID)), the Sec-WebSocket-Accept of RFC 6455 section 4.2.2.
fn websocket_accept_value(key: &str) -> String {
    let mut hashed_input = String::with_capacity(key.len() + WEBSOCKET_ACCEPT_GUID.len());
    hashed_input.push_str(key);
    hashed_input.push_str(WEBSOCKET_ACCEPT_GUID);
    base64_standard(&sha1_digest(hashed_input.as_bytes()))
}

/// SHA-1 (FIPS 180-4). Present because the accept value of the WebSocket
/// handshake is defined in terms of it; it is used for nothing else.
fn sha1_digest(message: &[u8]) -> [u8; 20] {
    let mut h: [u32; 5] = [0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476, 0xC3D2E1F0];

    let mut padded = message.to_vec();
    padded.push(0x80);
    while padded.len() % 64 != 56 {
        padded.push(0);
    }
    padded.extend_from_slice(&((message.len() as u64) * 8).to_be_bytes());

    for block in padded.chunks_exact(64) {
        let mut w = [0u32; 80];
        for (index, word) in block.chunks_exact(4).enumerate() {
            w[index] = u32::from_be_bytes([word[0], word[1], word[2], word[3]]);
        }
        for index in 16..80 {
            w[index] = (w[index - 3] ^ w[index - 8] ^ w[index - 14] ^ w[index - 16]).rotate_left(1);
        }

        let (mut a, mut b, mut c, mut d, mut e) = (h[0], h[1], h[2], h[3], h[4]);
        for (index, word) in w.iter().enumerate() {
            let (f, k) = match index {
                0..=19 => ((b & c) | ((!b) & d), 0x5A827999u32),
                20..=39 => (b ^ c ^ d, 0x6ED9EBA1),
                40..=59 => ((b & c) | (b & d) | (c & d), 0x8F1BBCDC),
                _ => (b ^ c ^ d, 0xCA62C1D6),
            };
            let temp = a
                .rotate_left(5)
                .wrapping_add(f)
                .wrapping_add(e)
                .wrapping_add(k)
                .wrapping_add(*word);
            e = d;
            d = c;
            c = b.rotate_left(30);
            b = a;
            a = temp;
        }
        h[0] = h[0].wrapping_add(a);
        h[1] = h[1].wrapping_add(b);
        h[2] = h[2].wrapping_add(c);
        h[3] = h[3].wrapping_add(d);
        h[4] = h[4].wrapping_add(e);
    }

    let mut digest = [0u8; 20];
    for (index, word) in h.iter().enumerate() {
        digest[index * 4..index * 4 + 4].copy_from_slice(&word.to_be_bytes());
    }
    digest
}

/// RFC 4648 standard base64 with padding.
fn base64_standard(bytes: &[u8]) -> String {
    const ALPHABET: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    let mut out = String::with_capacity((bytes.len() + 2) / 3 * 4);
    for chunk in bytes.chunks(3) {
        let b0 = chunk[0] as u32;
        let b1 = *chunk.get(1).unwrap_or(&0) as u32;
        let b2 = *chunk.get(2).unwrap_or(&0) as u32;
        let triple = (b0 << 16) | (b1 << 8) | b2;

        out.push(ALPHABET[(triple >> 18) as usize & 0x3F] as char);
        out.push(ALPHABET[(triple >> 12) as usize & 0x3F] as char);
        out.push(if chunk.len() > 1 {
            ALPHABET[(triple >> 6) as usize & 0x3F] as char
        } else {
            '='
        });
        out.push(if chunk.len() > 2 {
            ALPHABET[triple as usize & 0x3F] as char
        } else {
            '='
        });
    }
    out
}

#[cfg(test)]
mod rfc6455_handshake_vectors {
    use super::{base64_standard, sha1_digest, websocket_accept_value};

    #[test]
    fn sha1_matches_the_fips_180_4_example_digests() {
        assert_eq!(
            sha1_digest(b"abc")
                .iter()
                .map(|b| format!("{b:02x}"))
                .collect::<String>(),
            "a9993e364706816aba3e25717850c26c9cd0d89d"
        );
        assert_eq!(
            sha1_digest(b"")
                .iter()
                .map(|b| format!("{b:02x}"))
                .collect::<String>(),
            "da39a3ee5e6b4b0d3255bfef95601890afd80709"
        );
    }

    #[test]
    fn base64_matches_the_rfc_4648_test_vectors() {
        assert_eq!(base64_standard(b"f"), "Zg==");
        assert_eq!(base64_standard(b"fo"), "Zm8=");
        assert_eq!(base64_standard(b"foo"), "Zm9v");
        assert_eq!(base64_standard(b"foobar"), "Zm9vYmFy");
    }

    #[test]
    fn accept_value_matches_the_rfc_6455_section_1_3_example() {
        assert_eq!(
            websocket_accept_value("dGhlIHNhbXBsZSBub25jZQ=="),
            "s3pPLMBiTxaQ9kYGzzhZRbK+xOo="
        );
    }
}

File: hosts/python/transport/websocket_server.py
"""Quackoscope host (Python) -- transport layer.

One WebSocket endpoint at /ws (JSON control plane, binary data plane) plus the
built SPA served from ./dist as static files. Written on socket + threading from
the standard library; no third-party dependency. Nothing here imports opendaq.

The dist root is canonicalised once and every resolved file must sit under it,
so a request target such as "/../../Windows/win.ini" or "/../dist-evil/x" can
never escape the SPA build directory.
"""

import base64
import hashlib
import json
import mimetypes
import os
import socket
import struct
import threading
import traceback
from urllib.parse import unquote

from . import wire

_WEBSOCKET_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

_OPCODE_CONTINUATION = 0x0
_OPCODE_TEXT = 0x1
_OPCODE_BINARY = 0x2
_OPCODE_CLOSE = 0x8
_OPCODE_PING = 0x9
_OPCODE_PONG = 0xA


class WebSocketConnection:
    """One live socket. send_text and send_binary are safe from the reader
    thread and from any number of sample-pump threads at once."""

    def __init__(self, sock, peer):
        self._socket = sock
        self._send_lock = threading.Lock()
        self.peer = peer
        self.closed = False

    def send_text(self, text):
        self._send_frame(_OPCODE_TEXT, text.encode("utf-8"))

    def send_json(self, envelope):
        self.send_text(json.dumps(envelope, separators=(",", ":")))

    def send_binary(self, payload):
        self._send_frame(_OPCODE_BINARY, payload)

    def _send_frame(self, opcode, payload):
        length = len(payload)
        if length < 126:
            header = struct.pack("!BB", 0x80 | opcode, length)
        elif length < 65536:
            header = struct.pack("!BBH", 0x80 | opcode, 126, length)
        else:
            header = struct.pack("!BBQ", 0x80 | opcode, 127, length)
        with self._send_lock:
            if self.closed:
                return
            try:
                self._socket.sendall(header + payload)
            except OSError:
                self.closed = True

    def close(self):
        with self._send_lock:
            if self.closed:
                return
            self.closed = True
        try:
            self._socket.shutdown(socket.SHUT_RDWR)
        except OSError:
            pass
        try:
            self._socket.close()
        except OSError:
            pass


class _SocketReader:
    def __init__(self, sock):
        self._socket = sock
        self._buffer = b""

    def read_exactly(self, count):
        while len(self._buffer) < count:
            chunk = self._socket.recv(65536)
            if not chunk:
                raise ConnectionError("peer closed the socket")
            self._buffer += chunk
        out, self._buffer = self._buffer[:count], self._buffer[count:]
        return out

    def read_http_head(self):
        while b"\r\n\r\n" not in self._buffer:
            chunk = self._socket.recv(65536)
            if not chunk:
                raise ConnectionError("peer closed the socket before sending a request line")
            self._buffer += chunk
            if len(self._buffer) > 65536:
                raise ConnectionError("HTTP request head exceeded 64 KiB")
        head, self._buffer = self._buffer.split(b"\r\n\r\n", 1)
        return head.decode("latin-1")


def _read_websocket_message(reader):
    """Returns (opcode, payload) with continuation frames already joined."""
    message_opcode = None
    chunks = []
    while True:
        first, second = struct.unpack("!BB", reader.read_exactly(2))
        fin = bool(first & 0x80)
        opcode = first & 0x0F
        masked = bool(second & 0x80)
        length = second & 0x7F
        if length == 126:
            length = struct.unpack("!H", reader.read_exactly(2))[0]
        elif length == 127:
            length = struct.unpack("!Q", reader.read_exactly(8))[0]
        mask = reader.read_exactly(4) if masked else None
        payload = reader.read_exactly(length) if length else b""
        if mask:
            payload = bytes(b ^ mask[i % 4] for i, b in enumerate(payload))

        if opcode in (_OPCODE_CLOSE, _OPCODE_PING, _OPCODE_PONG):
            return opcode, payload

        if opcode != _OPCODE_CONTINUATION:
            message_opcode = opcode
        chunks.append(payload)
        if fin:
            return message_opcode, b"".join(chunks)


class HostServer:
    """Serves /ws and the SPA in ./dist. `handler` provides on_open(connection),
    on_close(connection) and on_request(connection, method, params), the last
    returning (ok, result, code, detail)."""

    def __init__(self, address, port, dist_dir, handler):
        self._address = address
        self._port = port
        self._handler = handler
        self._dist_dir = dist_dir
        self._canonical_dist_root = (
            os.path.realpath(dist_dir) if dist_dir and os.path.isdir(dist_dir) else None
        )
        self._listener = None

    def serve_forever(self):
        self._listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._listener.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._listener.bind((self._address, self._port))
        self._listener.listen(64)
        print(
            "[transport] listening on http://%s:%d and on the WebSocket endpoint ws://%s:%d/ws"
            % (self._address, self._port, self._address, self._port),
            flush=True,
        )
        if self._canonical_dist_root:
            print("[transport] serving the SPA build from %s" % self._canonical_dist_root, flush=True)
        else:
            print(
                '[transport] no SPA build at "%s": GET answers 503; the wire contract on /ws '
                "is still live" % self._dist_dir,
                flush=True,
            )
        while True:
            client, peer = self._listener.accept()
            threading.Thread(target=self._serve_one_socket, args=(client, peer), daemon=True).start()

    def _serve_one_socket(self, client, peer):
        peer_text = "%s:%d" % peer
        reader = _SocketReader(client)
        try:
            head = reader.read_http_head()
            request_line = head.split("\r\n", 1)[0]
            parts = request_line.split(" ")
            if len(parts) < 2:
                self._send_http(client, 400, "text/plain; charset=utf-8", b"malformed request line\n")
                return
            method, target = parts[0], parts[1]
            headers = {}
            for line in head.split("\r\n")[1:]:
                if ":" in line:
                    key, value = line.split(":", 1)
                    headers[key.strip().lower()] = value.strip()

            if target.split("?", 1)[0] == "/ws" and "websocket" in headers.get("upgrade", "").lower():
                self._accept_websocket(client, reader, headers, peer_text)
                return

            if method not in ("GET", "HEAD"):
                self._send_http(client, 405, "text/plain; charset=utf-8", b"only GET and HEAD are served\n")
                return
            self._serve_static_file(client, target, peer_text, body_wanted=(method == "GET"))
        except (ConnectionError, OSError):
            pass
        except Exception:
            traceback.print_exc()
        finally:
            try:
                client.close()
            except OSError:
                pass

    # --- WebSocket -------------------------------------------------------

    def _accept_websocket(self, client, reader, headers, peer_text):
        key = headers.get("sec-websocket-key")
        if not key:
            self._send_http(client, 400, "text/plain; charset=utf-8", b"missing Sec-WebSocket-Key\n")
            return
        accept = base64.b64encode(
            hashlib.sha1((key + _WEBSOCKET_GUID).encode("ascii")).digest()
        ).decode("ascii")
        client.sendall(
            (
                "HTTP/1.1 101 Switching Protocols\r\n"
                "Upgrade: websocket\r\n"
                "Connection: Upgrade\r\n"
                "Sec-WebSocket-Accept: " + accept + "\r\n\r\n"
            ).encode("ascii")
        )
        print("[transport] WebSocket /ws opened from %s" % peer_text, flush=True)

        connection = WebSocketConnection(client, peer_text)
        self._handler.on_open(connection)
        try:
            while True:
                opcode, payload = _read_websocket_message(reader)
                if opcode == _OPCODE_CLOSE:
                    break
                if opcode == _OPCODE_PING:
                    connection._send_frame(_OPCODE_PONG, payload)
                    continue
                if opcode == _OPCODE_PONG:
                    continue
                if opcode != _OPCODE_TEXT:
                    print(
                        "[transport] ignoring a %d-byte binary frame from %s: the control plane "
                        "is JSON text only" % (len(payload), peer_text),
                        flush=True,
                    )
                    continue
                self._answer_request(connection, payload.decode("utf-8", "replace"))
        except (ConnectionError, OSError, struct.error):
            pass
        finally:
            self._handler.on_close(connection)
            connection.close()
            print("[transport] WebSocket /ws from %s closed" % peer_text, flush=True)

    def _answer_request(self, connection, text):
        request, error_envelope = wire.decode_request(text)
        if request is None:
            connection.send_json(error_envelope)
            return
        ok, result, code, detail = self._handler.on_request(connection, request.method, request.params)
        if ok:
            connection.send_json(wire.encode_result(request.id, result))
        else:
            connection.send_json(wire.encode_error(request.id, code, detail))

    # --- static files ----------------------------------------------------

    def _serve_static_file(self, client, target, peer_text, body_wanted):
        if self._canonical_dist_root is None:
            self._send_http(
                client,
                503,
                "text/plain; charset=utf-8",
                (
                    'quackoscope-host-python: no SPA build at "%s".\nThe wire contract is still '
                    "live on /ws -- run the Vite dev server for the UI.\n" % self._dist_dir
                ).encode("utf-8"),
            )
            return

        path = unquote(target.split("?", 1)[0].split("#", 1)[0])
        if path in ("", "/"):
            path = "/index.html"

        # A decoded target is not usable as a location under the dist root: an
        # absolute path or a drive letter names a root of its own, and "\" is a
        # separator on Windows. Reject those outright, then canonicalise and
        # require the result to sit under the canonical dist root, so no sibling
        # directory such as "dist-evil" can ever pass as a child of "dist".
        if "\x00" in path or "\\" in path or ":" in path:
            self._refuse_static(client, target, peer_text, "it carries a path separator or a drive letter")
            return

        root = self._canonical_dist_root
        candidate = os.path.realpath(os.path.join(root, path.lstrip("/")))
        if os.path.normcase(candidate) != os.path.normcase(root) and not os.path.normcase(
            candidate
        ).startswith(os.path.normcase(root) + os.sep):
            print(
                "[transport] refused GET %s from %s: it resolves to %s, which is outside the "
                "dist root %s" % (target, peer_text, candidate, root),
                flush=True,
            )
            self._send_http(client, 403, "text/plain; charset=utf-8", b"outside the dist root\n")
            return

        # SPA fallback: an unknown route is index.html, an unknown asset is 404.
        if not os.path.isfile(candidate):
            if "." in os.path.basename(candidate):
                self._send_http(client, 404, "text/plain; charset=utf-8", b"no such file in dist\n")
                return
            candidate = os.path.join(root, "index.html")
            if not os.path.isfile(candidate):
                self._send_http(client, 404, "text/plain; charset=utf-8", b"no index.html in dist\n")
                return

        with open(candidate, "rb") as handle:
            body = handle.read()
        content_type = mimetypes.guess_type(candidate)[0] or "application/octet-stream"
        if content_type.startswith("text/") or content_type in (
            "application/javascript",
            "image/svg+xml",
        ):
            content_type += "; charset=utf-8"
        self._send_http(client, 200, content_type, body if body_wanted else b"", len(body))

    def _refuse_static(self, client, target, peer_text, refusal):
        print(
            "[transport] refused GET %s from %s: %s; dist root is %s"
            % (target, peer_text, refusal, self._canonical_dist_root),
            flush=True,
        )
        self._send_http(client, 400, "text/plain; charset=utf-8", b"unusable request target\n")

    @staticmethod
    def _send_http(client, status, content_type, body, content_length=None):
        reasons = {
            200: "OK",
            400: "Bad Request",
            403: "Forbidden",
            404: "Not Found",
            405: "Method Not Allowed",
            503: "Service Unavailable",
        }
        if content_length is None:
            content_length = len(body)
        head = (
            "HTTP/1.1 %d %s\r\nContent-Type: %s\r\nContent-Length: %d\r\n"
            "Cache-Control: no-store\r\nConnection: close\r\n\r\n"
            % (status, reasons.get(status, "Status"), content_type, content_length)
        )
        try:
            client.sendall(head.encode("latin-1") + body)
        except OSError:
            pass
